File: ios/Sources/MessengerNotificationsPlugin/FcmTokenRegistrar.swift
import Foundation

public enum FcmTokenRegistrar {
    /// Same `safe_storage` keys as Android (`FcmTokenRegistrar.java`).
    private static let baseUrlKeys = [
        "backendBaseUrl",
        "backend_url",
        "apiBaseUrl",
        "api_base_url",
        "serverUrl",
        "server_url"
    ]

    private static let defaultBaseUrl = "https://4.rw"
    private static let defaultFcmEndpointPath = "/api/users/fcm-token"
    private static let oneSignalEndpointPath = "/api/push/register"

    private static let requestTimeout: TimeInterval = 15

    private static func isTruthyStorageFlag(_ key: String) -> Bool {
        guard let raw = SafeStorageStore.get(key)?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() else {
            return false
        }
        return raw == "true" || raw == "1" || raw == "yes"
    }

    private static func resolvedJwt() -> String? {
        let jwt = SafeStorageStore.get("token") ?? SafeStorageStore.get("authToken")
        guard let jwtToken = jwt, !jwtToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        return jwtToken
    }

    private static func resolvedBaseUrl() -> String {
        for key in baseUrlKeys {
            if let value = SafeStorageStore.get(key), !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                return value
            }
        }
        return defaultBaseUrl
    }

    private static func registrationURL(base: String, path: String) -> String? {
        let normalizedBase = base.hasSuffix("/") ? String(base.dropLast()) : base
        let normalizedPath = path.hasPrefix("/") ? path : "/" + path
        return normalizedBase + normalizedPath
    }

    /// Stores an FCM registration token (e.g. from Firebase on iOS) in `safe_storage` and attempts backend registration.
    /// Clears `fcmTokenRegistered` when the token changes so the server is updated.
    public static func updateFcmToken(_ token: String?) {
        let normalized = token?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        if normalized.isEmpty {
            SafeStorageStore.remove("fcmToken")
            SafeStorageStore.set("fcmTokenRegistered", value: "false")
            print("🔔 [iOS/FcmTokenRegistrar] Cleared FCM token from local storage.")
            return
        }

        let existing = SafeStorageStore.get("fcmToken")?.trimmingCharacters(in: .whitespacesAndNewlines)
        SafeStorageStore.set("fcmToken", value: normalized)
        SafeStorageStore.set("pushTokenType", value: "fcm")
        SafeStorageStore.set("pushTokenUpdatedAt", value: "\(Int(Date().timeIntervalSince1970 * 1000))")

        if existing != normalized {
            SafeStorageStore.set("fcmTokenRegistered", value: "false")
            print("🔔 [iOS/FcmTokenRegistrar] Stored new FCM token (length=\(normalized.count)).")
        }

        registerIfPossible()
    }

    public static func updateOneSignalPlayerId(_ playerId: String?) {
        let normalizedPlayerId = playerId?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        if normalizedPlayerId.isEmpty {
            SafeStorageStore.remove("onesignalPlayerId")
            SafeStorageStore.remove("onesignalSubscriptionId")
            SafeStorageStore.set("onesignalPlayerIdRegistered", value: "false")
            print("🔔 [iOS/FcmTokenRegistrar] Cleared OneSignal subscription id from local storage.")
            return
        }

        let existingPlayerId = SafeStorageStore.get("onesignalPlayerId")?.trimmingCharacters(in: .whitespacesAndNewlines)
        SafeStorageStore.set("onesignalPlayerId", value: normalizedPlayerId)
        SafeStorageStore.set("onesignalSubscriptionId", value: normalizedPlayerId)
        SafeStorageStore.set("pushTokenType", value: "onesignal_player_id")
        SafeStorageStore.set("pushTokenUpdatedAt", value: "\(Int(Date().timeIntervalSince1970 * 1000))")

        if existingPlayerId != normalizedPlayerId {
            SafeStorageStore.set("onesignalPlayerIdRegistered", value: "false")
            print("🔔 [iOS/FcmTokenRegistrar] Stored new OneSignal subscription id: \(normalizedPlayerId)")
        }

        registerIfPossible()
    }

    /// Registers with the backend when possible: **FCM** (`/api/users/fcm-token` or `fcmTokenEndpoint`) and/or **OneSignal** (`/api/push/register`), using the same `safe_storage` keys as Android.
    public static func registerIfPossible() {
        registerFcmTokenIfPossible()
        registerOneSignalIfPossible()
    }

    private static func registerFcmTokenIfPossible() {
        guard let fcmToken = SafeStorageStore.get("fcmToken")?.trimmingCharacters(in: .whitespacesAndNewlines), !fcmToken.isEmpty else {
            print("🔔 [iOS/FcmTokenRegistrar] No FCM token stored, skipping FCM registration.")
            return
        }

        if isTruthyStorageFlag("fcmTokenRegistered") {
            print("🔔 [iOS/FcmTokenRegistrar] FCM token already registered with server.")
            return
        }

        guard let jwtToken = resolvedJwt() else {
            print("🔔 [iOS/FcmTokenRegistrar] No JWT token yet (user not logged in); FCM registration deferred.")
            return
        }

        let base = resolvedBaseUrl()
        var path = SafeStorageStore.get("fcmTokenEndpoint")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if path.isEmpty {
            path = defaultFcmEndpointPath
        }

        guard let urlString = registrationURL(base: base, path: path), let url = URL(string: urlString) else {
            print("🔔 [iOS/FcmTokenRegistrar] Invalid FCM registration URL.")
            return
        }

        let payload: [String: Any] = ["fcmToken": fcmToken]
        guard let body = try? JSONSerialization.data(withJSONObject: payload, options: []) else {
            print("🔔 [iOS/FcmTokenRegistrar] Failed to encode FCM registration payload.")
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = requestTimeout
        request.setValue("Bearer \(jwtToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = body

        print("🔔 [iOS/FcmTokenRegistrar] Dispatching FCM token registration to \(urlString)")

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error {
                print("🔔 [iOS/FcmTokenRegistrar] FCM registration error: \(error.localizedDescription)")
                return
            }
            guard let http = response as? HTTPURLResponse else {
                print("🔔 [iOS/FcmTokenRegistrar] FCM registration failed: No HTTP response")
                return
            }
            if 200 ..< 300 ~= http.statusCode {
                print("🔔 [iOS/FcmTokenRegistrar] FCM token registered successfully (HTTP \(http.statusCode)).")
                SafeStorageStore.set("fcmTokenRegistered", value: "true")
            } else {
                let responseBody = String(data: data ?? Data(), encoding: .utf8) ?? ""
                print("🔔 [iOS/FcmTokenRegistrar] FCM registration failed (HTTP \(http.statusCode)) body=\(responseBody)")
            }
        }.resume()
    }

    private static func registerOneSignalIfPossible() {
        guard let playerId = SafeStorageStore.get("onesignalPlayerId")?.trimmingCharacters(in: .whitespacesAndNewlines), !playerId.isEmpty else {
            print("🔔 [iOS/FcmTokenRegistrar] No OneSignal subscription id stored, skipping OneSignal registration.")
            return
        }

        if isTruthyStorageFlag("onesignalPlayerIdRegistered") {
            print("🔔 [iOS/FcmTokenRegistrar] OneSignal subscription id already registered with server.")
            return
        }

        guard let jwtToken = resolvedJwt() else {
            print("🔔 [iOS/FcmTokenRegistrar] No JWT token yet (user not logged in); OneSignal registration deferred.")
            return
        }

        let base = resolvedBaseUrl()
        guard let urlString = registrationURL(base: base, path: oneSignalEndpointPath), let url = URL(string: urlString) else {
            print("🔔 [iOS/FcmTokenRegistrar] Invalid OneSignal registration URL.")
            return
        }

        let payload: [String: Any] = [
            "playerId": playerId,
            "platform": "ios"
        ]

        guard let body = try? JSONSerialization.data(withJSONObject: payload, options: []) else {
            print("🔔 [iOS/FcmTokenRegistrar] Failed to encode OneSignal registration payload.")
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = requestTimeout
        request.setValue("Bearer \(jwtToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = body

        print("🔔 [iOS/FcmTokenRegistrar] Dispatching OneSignal subscription registration to \(urlString)")

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error {
                print("🔔 [iOS/FcmTokenRegistrar] OneSignal registration error: \(error.localizedDescription)")
                return
            }
            guard let http = response as? HTTPURLResponse else {
                print("🔔 [iOS/FcmTokenRegistrar] OneSignal registration failed: No HTTP response")
                return
            }
            if 200 ..< 300 ~= http.statusCode {
                print("🔔 [iOS/FcmTokenRegistrar] OneSignal subscription id registered successfully (HTTP \(http.statusCode)).")
                SafeStorageStore.set("onesignalPlayerIdRegistered", value: "true")
            } else {
                let responseBody = String(data: data ?? Data(), encoding: .utf8) ?? ""
                print("🔔 [iOS/FcmTokenRegistrar] OneSignal registration failed (HTTP \(http.statusCode)) body=\(responseBody)")
            }
        }.resume()
    }
}

File: android/src/main/java/com/codecraft_studio/messenger/notifications/FcmTokenRegistrar.java
package com.codecraft_studio.messenger.notifications;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;
import android.util.Log;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Registers the device's FCM token with the backend API.
 *
 * SharedPreferences keys used (all in "safe_storage"):
 *   fcmToken            – the raw FCM registration token
 *   fcmTokenRegistered  – boolean; true once ACKed by the server
 *   token / authToken   – the user's JWT
 *   backendBaseUrl / backend_url / apiBaseUrl / api_base_url / serverUrl / server_url
 *                       – base URL of the backend
 *   fcmTokenEndpoint    – optional override for the registration endpoint path
 */
final class FcmTokenRegistrar {

    private static final String TAG = "FcmTokenRegistrar";
    private static final String PREFS_NAME = "safe_storage";
    private static final String DEFAULT_ENDPOINT_PATH = "/api/users/fcm-token";
    private static final int CONNECT_TIMEOUT_MS = 8_000;
    private static final int READ_TIMEOUT_MS = 10_000;

    private static final List<String> BASE_URL_KEYS = Arrays.asList(
        "backendBaseUrl",
        "backend_url",
        "apiBaseUrl",
        "api_base_url",
        "serverUrl",
        "server_url"
    );

    private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();

    private FcmTokenRegistrar() {}

    static void registerIfPossible(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);

        String fcmToken = prefs.getString("fcmToken", null);
        if (TextUtils.isEmpty(fcmToken)) {
            Log.d(TAG, "No FCM token stored, skipping registration.");
            return;
        }

        boolean alreadyRegistered = prefs.getBoolean("fcmTokenRegistered", false);
        if (alreadyRegistered) {
            Log.d(TAG, "FCM token already registered with server.");
            return;
        }

        String jwtToken = firstNonEmpty(prefs.getString("token", null), prefs.getString("authToken", null));
        if (TextUtils.isEmpty(jwtToken)) {
            Log.d(TAG, "No JWT token yet (user not logged in). JS will register token after login.");
            return;
        }

        String baseUrl = null;
        for (String key : BASE_URL_KEYS) {
            String v = prefs.getString(key, null);
            if (!TextUtils.isEmpty(v)) {
                baseUrl = v;
                break;
            }
        }
        if (TextUtils.isEmpty(baseUrl)) {
            Log.w(TAG, "No backend base URL configured, cannot register FCM token.");
            return;
        }

        final String finalFcmToken = fcmToken;
        final String finalJwtToken = jwtToken;
        final String registrationUrl = buildUrl(prefs, baseUrl);
        final Context appContext = context.getApplicationContext();

        Log.i(TAG, "Dispatching FCM token registration to " + registrationUrl);
        EXECUTOR.execute(() -> doRegister(appContext, finalFcmToken, finalJwtToken, registrationUrl));
    }

    private static void doRegister(Context context, String fcmToken, String jwtToken, String url) {
        HttpURLConnection conn = null;
        try {
            byte[] body = ("{\"fcmToken\":\"" + fcmToken + "\"}").getBytes(StandardCharsets.UTF_8);

            conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setRequestMethod("POST");
            conn.setConnectTimeout(CONNECT_TIMEOUT_MS);
            conn.setReadTimeout(READ_TIMEOUT_MS);
            conn.setDoOutput(true);
            conn.setRequestProperty("Authorization", "Bearer " + jwtToken);
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setRequestProperty("Accept", "application/json");
            conn.setFixedLengthStreamingMode(body.length);

            try (OutputStream os = conn.getOutputStream()) {
                os.write(body);
            }

            int status = conn.getResponseCode();
            if (status >= 200 && status < 300) {
                Log.i(TAG, "FCM token registered successfully (HTTP " + status + ").");
                context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit().putBoolean("fcmTokenRegistered", true).apply();
            } else {
                Log.w(TAG, "FCM token registration failed (HTTP " + status + "). Will retry.");
            }
        } catch (Exception e) {
            Log.w(TAG, "FCM token registration error – will retry next time.", e);
        } finally {
            if (conn != null) {
                conn.disconnect();
            }
        }
    }

    private static String buildUrl(SharedPreferences prefs, String baseUrl) {
        String endpointPath = prefs.getString("fcmTokenEndpoint", DEFAULT_ENDPOINT_PATH);
        if (TextUtils.isEmpty(endpointPath)) {
            endpointPath = DEFAULT_ENDPOINT_PATH;
        }
        String normalizedBase = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        String normalizedPath = endpointPath.startsWith("/") ? endpointPath : "/" + endpointPath;
        return normalizedBase + normalizedPath;
    }

    private static String firstNonEmpty(String... values) {
        for (String v : values) {
            if (!TextUtils.isEmpty(v)) return v;
        }
        return null;
    }
}
